File: codes/cpp/chap08/answer_A51.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int Q;
int QueryType[100009]; string x[100009];
stack<string> S;

int main() {
	// 入力
	cin >> Q;
	for (int i = 1; i <= Q; i++) {
		cin >> QueryType[i];
		if (QueryType[i] == 1) cin >> x[i];
	}

	// クエリの処理
	for (int i = 1; i <= Q; i++) {
		if (QueryType[i] == 1) S.push(x[i]);
		if (QueryType[i] == 2) cout << S.top() << endl;
		if (QueryType[i] == 3) S.pop();
	}
	return 0;
}

File: codes/java/chap08/answer_A51.java
import java.util.*;

class Main {
	public static void main(String[] args) {
		// 入力
		Scanner sc = new Scanner(System.in);
		int Q = sc.nextInt();
		int[] queryType = new int[Q + 1];
		String[] x = new String[Q + 1];
		for (int i = 1; i <= Q; i++) {
			queryType[i] = sc.nextInt();
			if (queryType[i] == 1) {
				x[i] = sc.next();
			}
		}

		// クエリの処理
		Stack<String> S = new Stack<>();
		for (int i = 1; i <= Q; i++) {
			if (queryType[i] == 1) {
				S.push(x[i]);
			}
			if (queryType[i] == 2) {
				System.out.println(S.peek());
			}
			if (queryType[i] == 3) {
				S.pop();
			}
		}
	}
}